File: 111-二叉树的最小深度.cpp
class Solution {
public:
    int minDepth(TreeNode* root) {
        // BFS 非递归
        if(!root)
            return 0;

        int result = INT_MAX;
        queue<pair<TreeNode*, int>> myqueue;
        myqueue.push(make_pair(root, 1));
        TreeNode* node;
        int depth;

        while(!myqueue.empty()){
            node = myqueue.front().first;
            depth = myqueue.front().second;
            myqueue.pop();
            if(!node)
                continue;
            else if(!node->left && !node->right)
                result = min(result, depth);
            else{
                myqueue.push(make_pair(node->left, depth + 1));
                myqueue.push(make_pair(node->right, depth + 1));
            }
        }

        return result;
    }
	};

File: 111-二叉树的最小深度.py
class Solution:
    def minDepth(self, root: TreeNode) -> int:
        ## DFS - 非递归
        if not root:
            return 0
            
        result = float('inf')
        mystack = [(root, 1)]

        while mystack:
            node, depth = mystack.pop()
            if not node:
                continue
            elif not node.left and not node.right:
                result = min(result, depth)
            else:
                mystack.append((node.left, depth + 1))
                mystack.append((node.right, depth + 1))
        
        return result
